### PR TITLE
Add Arm Inertia

### PR DESCRIPTION
**Pull request by kilbith**
Addresses https://github.com/minetest/minetest/issues/5174
Demo (outdated): https://www.youtube.com/watch?v=ls9vxBTr0hg

Optional feature, enabled by default. Wielding a pickaxe / sword gives the best effect.

Some values may need to be tweaked though.

### CODE REVIEW

File: src/camera.h [hunk 1]
```diff
@@ -188,6 +188,12 @@ class Camera
 	// Camera offset
 	v3s16 m_camera_offset;
 
+	v3f m_wieldmesh_offset  = v3f(55.0f, -35.0f, 65.0f);
+	float m_timer[2]        = {0.0f};
```
Comment (nerzhul) [inline src/camera.h]: please init both offsets = { 0.0f, 0.0f }

**Review by nerzhul — CHANGES_REQUESTED** [review]

**Review by nerzhul — APPROVED** [review]

File: src/camera.h [hunk 2]
```diff
@@ -188,6 +188,12 @@ class Camera
 	// Camera offset
 	v3s16 m_camera_offset;
 
+	v3f m_wieldmesh_offset  = v3f(55.0f, -35.0f, 65.0f);
+	float m_timer[2]        = { 0.0f, 0.0f };
+	float m_cam_vel[2]      = { 0.0f, 0.0f };
+	float m_cam_vel_old[2]  = { 0.0f, 0.0f };
+	float m_last_cam_pos[2] = { 0.0f, 0.0f };
```
Comment (red-001) [inline src/camera.h]: Can you use `v2f`?

**Review by nerzhul — APPROVED** [review]

File: builtin/settingtypes.txt [hunk 3]
```diff
@@ -469,6 +469,9 @@ enable_waving_plants (Waving plants) bool false
 
 [***Advanced]
 
+#    Arm inertia
```
Comment (sfan5) [inline builtin/settingtypes.txt]: longer description would be nice, something like
`more realistic movement of arm when turning view`

**Comment by paramat** [comment]
:-1: Sorry this is unrealistic and belongs in a photo-realistic FPS.
Or at least disabled by default due to it's fanciness.
I thought you were not goimg to backport anything from your fork to to MT?

**Comment by nerzhul** [comment]
it's more realistic than you outside dungeons which are ugly. This pr permits to have a nice game, unlike mapgen v7 and your outside dungeons.

**Comment by Fixer-007** [comment]
You can always make it configurable so mod or subgame can choose realistic or nonrealistic way.

**Comment by paramat** [comment]
>  This pr permits to have a nice game, unlike mapgen v7 and your outside dungeons.

Mgv7 is much liked by many, and many like the protruding dungeons, but many do not also of course.

> also your stupidity is another time shown

Personal insult again, please stop.

> When do we refuse backports from forks ?

kilbith stated he would not backport from his fork, so i was wondering.

>  it's better than hidding bugs as features like you do.

It is an intentional feature.

**Review by nerzhul — CHANGES_REQUESTED** [review]
Hmmm i tested it and strangely when enabled i don't have any arm anymore

**Comment by kilbith** [comment]
Please respect my PR and don't hijack it with drama and offtopic.

**Comment by celeron55** [comment]
Seems harmless and fine to me.

If players complain, it can be straightforwardly tweaked or removed.

File: src/camera.cpp [hunk 4]
```diff
@@ -380,12 +381,85 @@ void Camera::update(LocalPlayer* player, f32 frametime, f32 busytime,
 	m_cameranode->setAspectRatio(m_aspect);
 	m_cameranode->setFOV(m_fov_y);
 
-	float wieldmesh_offset_Y = -35 + player->getPitch() * 0.05;
-	wieldmesh_offset_Y = rangelim(wieldmesh_offset_Y, -52, -32);
+	if (m_arm_inertia) {
```
Comment (nerzhul) [inline src/camera.cpp]: is this possible to extract all code in this condition to a function please ?

**Review by nerzhul — APPROVED** [review]

**Comment by nerzhul** [comment]
Strangely on a pure clean build it works (cmake workdir removed). If possible having a setting to manage the speed of the movement can be nice

File: src/camera.cpp [hunk 5]
```diff
@@ -193,6 +194,82 @@ void Camera::step(f32 dtime)
 	}
 }
 
+void Camera::add_arm_inertia(const float &player_yaw, const f32 &frametime)
```
Comment (nerzhul) [inline src/camera.cpp]: please pass float as value 4b < 8b (ref 64bits)

**Comment by bigfoot547** [comment]
> :-1: Sorry this is **unrealistic** and i feel it belongs in a **photo-realistic** FPS.

Or maybe a photo-_un_-realistic FPS. :wink:

**Comment by SmallJoker** [comment]
I just tested the patch and it looks very nice. The only thing I could complain is the abrupt stop when the item got back on place, but it looks good even without an extra smooth or wiggle effect there. KISS. :+1: once the comment is handled.

File: src/camera.cpp [hunk 6]
```diff
@@ -193,6 +194,82 @@ void Camera::step(f32 dtime)
 	}
 }
 
+void Camera::add_arm_inertia(f32 player_yaw, f32 frametime)
```
Comment (SmallJoker) [inline src/camera.cpp]: `addArmInertia` to follow our function naming rules

**Comment by paramat** [comment]
Removed my -1. I've been watching MC videos and noticed that with a moderate displacement, smooth motion and fast return it doesn't look bad.
I was having a day of arguments as you know and that affected my reaction to this PR, sorry for that.

**Pull request merged**